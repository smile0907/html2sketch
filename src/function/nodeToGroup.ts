/* eslint-disable no-console */
import nodeToLayers from './nodeToLayers';
import { isNodeVisible } from '../utils/visibility';
import { getChildNodeList } from '../utils/hierarchy';
import { getName } from '../utils/name';
import { Group, Style, AnyLayer } from '../model';
import { isExistPseudoText, isExistPseudoShape } from '../utils/shape';

export interface Options {
  postTransform?: (group: AnyLayer) => AnyLayer;
  getGroupName: (node: Element) => string;
}
/**
 * 将一个节点和其包含的所有子级转为 Group 对象
 * @param node
 * @param options
 */
const nodeToGroup = (node: Element, options?: Options): AnyLayer => {
  if (!node) throw Error('解析对象不存在 请检查传入对象');

  const bcr = node.getBoundingClientRect();
  const { left, top } = bcr;
  const width = bcr.right - bcr.left;
  const height = bcr.bottom - bcr.top;

  const layers = nodeToLayers(node) || [];

  // ---------- 处理父节点 ------ //
  if (node.nodeName !== 'svg') {
    const childNodeList = getChildNodeList(node);

    // Recursively collect child groups for child nodes
    childNodeList.forEach((childNode) => {
      layers.push(nodeToGroup(childNode, options));

      // Traverse the shadow DOM if present
      if (childNode.shadowRoot) {
        Array.from(childNode.shadowRoot.children)
          .filter((n) => isNodeVisible(n))
          .map((n) => nodeToGroup(n))
          .forEach((layer) => layers.push(layer));
      }
    });
  }

  // Now build a group for all these children

  const styles = getComputedStyle(node);
  const { opacity } = styles;

  const group = new Group({ x: left, y: top, width, height });
  const groupStyle = new Style();

  groupStyle.opacity = opacity;
  group.style = groupStyle;

  layers
    .filter((l) => l)
    .forEach((layer) => {
      group.addLayer(layer);
    });

  // 处理多余的层级
  if (
    group.layers.length === 1 &&
    (group.layers[0].class === 'rectangle' ||
      group.layers[0].class === 'text' ||
      group.layers[0].class === 'svg' ||
      group.layers[0].class === 'group')
  ) {
    const layer = group.layers[0];
    console.log(
      `[nodeToSketchGroup]该 group 只包含一个子级 [${layer.class}]: ${layer.name} ,丢弃...`,
    );
    // 将父级的图层关系还给子集
    layer.x += group.x;
    layer.y += group.y;
    return layer;
  }

  if (
    group.layers.length === 0 &&
    !isExistPseudoText(node) &&
    !isExistPseudoShape(node)
  ) {
    console.log('[nodeToSketchGroup]该 group 是空的,丢弃...');
    // @ts-ignore
    // eslint-disable-next-line consistent-return
    return;
  }

  if (options && options.getGroupName) {
    group.name = options.getGroupName(node);
  } else {
    group.name = getName(node.nodeName);
  }

  group.className = node.className;

  return group;
};

export default nodeToGroup;
