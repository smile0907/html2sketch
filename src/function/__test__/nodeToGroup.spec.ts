import nodeToGroup from '../nodeToGroup';

describe('单个', function () {
  afterEach(() => {
    document.body.innerHTML = '';
  });
  beforeEach(() => {
    document.body.innerHTML = `<div id="group">123</div>`;
  });
  it('正常解析', function () {
    const node = document.getElementById('group') as HTMLDivElement;
    const group = nodeToGroup(node);
    expect(group.toSketchJSON()).toMatchSnapshot();
  });
});

describe('多个解析', function () {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  beforeEach(() => {
    document.body.innerHTML = `<div id="group"><div>123</div><div>13333</div></div>`;
  });
  it('多个解析', () => {
    const node = document.getElementById('group') as HTMLDivElement;
    const group = nodeToGroup(node);
    expect(group.toSketchJSON()).toMatchSnapshot();
  });
});
