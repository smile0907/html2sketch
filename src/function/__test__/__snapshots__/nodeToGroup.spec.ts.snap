// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`单个 正常解析 1`] = `
Object {
  "_class": "text",
  "attributedString": Object {
    "_class": "attributedString",
    "attributes": Array [
      Object {
        "_class": "stringAttribute",
        "attributes": Object {
          "MSAttributedStringColorAttribute": Object {
            "_class": "color",
            "alpha": 1,
            "blue": 0,
            "green": 0,
            "red": 0,
          },
          "MSAttributedStringFontAttribute": Object {
            "_class": "fontDescriptor",
            "attributes": Object {
              "name": "PingFangSC-Regular",
              "size": 16,
            },
          },
          "MSAttributedStringTextTransformAttribute": 0,
          "kerning": 0,
          "paragraphStyle": Object {
            "_class": "paragraphStyle",
            "alignment": 0,
            "maximumLineHeight": 22,
            "minimumLineHeight": 22,
          },
          "strikethroughStyle": 0,
          "underlineStyle": 0,
        },
        "length": 3,
        "location": 0,
      },
    ],
    "string": "123",
  },
  "automaticallyDrawOnUnderlyingPath": false,
  "booleanOperation": -1,
  "clippingMaskMode": 0,
  "do_objectID": "UUID",
  "dontSynchroniseWithSymbol": false,
  "exportOptions": Object {
    "_class": "exportOptions",
    "exportFormats": Array [],
    "includedLayerIds": Array [],
    "layerOptions": 0,
    "shouldTrim": false,
  },
  "frame": Object {
    "_class": "rect",
    "constrainProportions": false,
    "height": 22,
    "width": 25.609375,
    "x": 8,
    "y": 8,
  },
  "glyphBounds": "",
  "hasClippingMask": false,
  "isFixedToViewport": false,
  "isFlippedHorizontal": false,
  "isFlippedVertical": false,
  "isLocked": false,
  "isVisible": true,
  "layerListExpandedType": 0,
  "lineSpacingBehaviour": 2,
  "name": "123",
  "nameIsFixed": false,
  "resizingConstraint": 63,
  "resizingType": 0,
  "rotation": 0,
  "shouldBreakMaskChain": false,
  "style": Object {
    "_class": "style",
    "borderOptions": Object {
      "_class": "borderOptions",
      "dashPattern": Array [],
      "isEnabled": true,
      "lineCapStyle": 0,
      "lineJoinStyle": 0,
    },
    "borders": Array [],
    "colorControls": Object {
      "_class": "colorControls",
      "brightness": 0,
      "contrast": 1,
      "hue": 0,
      "isEnabled": false,
      "saturation": 1,
    },
    "contextSettings": Object {
      "_class": "graphicsContextSettings",
      "blendMode": 0,
      "opacity": 1,
    },
    "do_objectID": "UUID",
    "endMarkerType": 0,
    "fills": Array [],
    "innerShadows": Array [],
    "miterLimit": 10,
    "shadows": Array [],
    "startMarkerType": 0,
    "windingRule": 1,
  },
  "textBehaviour": 0,
  "userInfo": undefined,
}
`;

exports[`多个解析 多个解析 1`] = `
Object {
  "_class": "group",
  "booleanOperation": -1,
  "clippingMaskMode": 0,
  "do_objectID": "UUID",
  "exportOptions": Object {
    "_class": "exportOptions",
    "exportFormats": Array [],
    "includedLayerIds": Array [],
    "layerOptions": 0,
    "shouldTrim": false,
  },
  "frame": Object {
    "_class": "rect",
    "constrainProportions": false,
    "height": 44,
    "width": 1008,
    "x": 8,
    "y": 8,
  },
  "groupLayout": Object {
    "_class": "MSImmutableFreeformGroupLayout",
  },
  "hasClickThrough": false,
  "hasClippingMask": false,
  "isFixedToViewport": false,
  "isFlippedHorizontal": false,
  "isFlippedVertical": false,
  "isLocked": false,
  "isVisible": true,
  "layerListExpandedType": 0,
  "layers": Array [
    Object {
      "_class": "text",
      "attributedString": Object {
        "_class": "attributedString",
        "attributes": Array [
          Object {
            "_class": "stringAttribute",
            "attributes": Object {
              "MSAttributedStringColorAttribute": Object {
                "_class": "color",
                "alpha": 1,
                "blue": 0,
                "green": 0,
                "red": 0,
              },
              "MSAttributedStringFontAttribute": Object {
                "_class": "fontDescriptor",
                "attributes": Object {
                  "name": "PingFangSC-Regular",
                  "size": 16,
                },
              },
              "MSAttributedStringTextTransformAttribute": 0,
              "kerning": 0,
              "paragraphStyle": Object {
                "_class": "paragraphStyle",
                "alignment": 0,
                "maximumLineHeight": 22,
                "minimumLineHeight": 22,
              },
              "strikethroughStyle": 0,
              "underlineStyle": 0,
            },
            "length": 3,
            "location": 0,
          },
        ],
        "string": "123",
      },
      "automaticallyDrawOnUnderlyingPath": false,
      "booleanOperation": -1,
      "clippingMaskMode": 0,
      "do_objectID": "UUID",
      "dontSynchroniseWithSymbol": false,
      "exportOptions": Object {
        "_class": "exportOptions",
        "exportFormats": Array [],
        "includedLayerIds": Array [],
        "layerOptions": 0,
        "shouldTrim": false,
      },
      "frame": Object {
        "_class": "rect",
        "constrainProportions": false,
        "height": 22,
        "width": 25.609375,
        "x": 0,
        "y": 0,
      },
      "glyphBounds": "",
      "hasClippingMask": false,
      "isFixedToViewport": false,
      "isFlippedHorizontal": false,
      "isFlippedVertical": false,
      "isLocked": false,
      "isVisible": true,
      "layerListExpandedType": 0,
      "lineSpacingBehaviour": 2,
      "name": "123",
      "nameIsFixed": false,
      "resizingConstraint": 63,
      "resizingType": 0,
      "rotation": 0,
      "shouldBreakMaskChain": false,
      "style": Object {
        "_class": "style",
        "borderOptions": Object {
          "_class": "borderOptions",
          "dashPattern": Array [],
          "isEnabled": true,
          "lineCapStyle": 0,
          "lineJoinStyle": 0,
        },
        "borders": Array [],
        "colorControls": Object {
          "_class": "colorControls",
          "brightness": 0,
          "contrast": 1,
          "hue": 0,
          "isEnabled": false,
          "saturation": 1,
        },
        "contextSettings": Object {
          "_class": "graphicsContextSettings",
          "blendMode": 0,
          "opacity": 1,
        },
        "do_objectID": "UUID",
        "endMarkerType": 0,
        "fills": Array [],
        "innerShadows": Array [],
        "miterLimit": 10,
        "shadows": Array [],
        "startMarkerType": 0,
        "windingRule": 1,
      },
      "textBehaviour": 0,
      "userInfo": undefined,
    },
    Object {
      "_class": "text",
      "attributedString": Object {
        "_class": "attributedString",
        "attributes": Array [
          Object {
            "_class": "stringAttribute",
            "attributes": Object {
              "MSAttributedStringColorAttribute": Object {
                "_class": "color",
                "alpha": 1,
                "blue": 0,
                "green": 0,
                "red": 0,
              },
              "MSAttributedStringFontAttribute": Object {
                "_class": "fontDescriptor",
                "attributes": Object {
                  "name": "PingFangSC-Regular",
                  "size": 16,
                },
              },
              "MSAttributedStringTextTransformAttribute": 0,
              "kerning": 0,
              "paragraphStyle": Object {
                "_class": "paragraphStyle",
                "alignment": 0,
                "maximumLineHeight": 22,
                "minimumLineHeight": 22,
              },
              "strikethroughStyle": 0,
              "underlineStyle": 0,
            },
            "length": 5,
            "location": 0,
          },
        ],
        "string": "13333",
      },
      "automaticallyDrawOnUnderlyingPath": false,
      "booleanOperation": -1,
      "clippingMaskMode": 0,
      "do_objectID": "UUID",
      "dontSynchroniseWithSymbol": false,
      "exportOptions": Object {
        "_class": "exportOptions",
        "exportFormats": Array [],
        "includedLayerIds": Array [],
        "layerOptions": 0,
        "shouldTrim": false,
      },
      "frame": Object {
        "_class": "rect",
        "constrainProportions": false,
        "height": 22,
        "width": 44.8125,
        "x": 0,
        "y": 22,
      },
      "glyphBounds": "",
      "hasClippingMask": false,
      "isFixedToViewport": false,
      "isFlippedHorizontal": false,
      "isFlippedVertical": false,
      "isLocked": false,
      "isVisible": true,
      "layerListExpandedType": 0,
      "lineSpacingBehaviour": 2,
      "name": "13333",
      "nameIsFixed": false,
      "resizingConstraint": 63,
      "resizingType": 0,
      "rotation": 0,
      "shouldBreakMaskChain": false,
      "style": Object {
        "_class": "style",
        "borderOptions": Object {
          "_class": "borderOptions",
          "dashPattern": Array [],
          "isEnabled": true,
          "lineCapStyle": 0,
          "lineJoinStyle": 0,
        },
        "borders": Array [],
        "colorControls": Object {
          "_class": "colorControls",
          "brightness": 0,
          "contrast": 1,
          "hue": 0,
          "isEnabled": false,
          "saturation": 1,
        },
        "contextSettings": Object {
          "_class": "graphicsContextSettings",
          "blendMode": 0,
          "opacity": 1,
        },
        "do_objectID": "UUID",
        "endMarkerType": 0,
        "fills": Array [],
        "innerShadows": Array [],
        "miterLimit": 10,
        "shadows": Array [],
        "startMarkerType": 0,
        "windingRule": 1,
      },
      "textBehaviour": 0,
      "userInfo": undefined,
    },
  ],
  "name": "编组",
  "nameIsFixed": false,
  "resizingConstraint": 63,
  "resizingType": 0,
  "rotation": 0,
  "shouldBreakMaskChain": false,
  "style": Object {
    "_class": "style",
    "borderOptions": Object {
      "_class": "borderOptions",
      "dashPattern": Array [],
      "isEnabled": true,
      "lineCapStyle": 0,
      "lineJoinStyle": 0,
    },
    "borders": Array [],
    "colorControls": Object {
      "_class": "colorControls",
      "brightness": 0,
      "contrast": 1,
      "hue": 0,
      "isEnabled": false,
      "saturation": 1,
    },
    "contextSettings": Object {
      "_class": "graphicsContextSettings",
      "blendMode": 0,
      "opacity": 1,
    },
    "do_objectID": "UUID",
    "endMarkerType": 0,
    "fills": Array [],
    "innerShadows": Array [],
    "miterLimit": 10,
    "shadows": Array [],
    "startMarkerType": 0,
    "windingRule": 1,
  },
  "userInfo": null,
}
`;
